src/guiengine/event_handler.cpp

```diff
@@ -427,172 +394,243 @@ const bool NAVIGATION_DEBUG = false;
-        for (int n = 1; n < 10 && !found; n++)
+        Widget* closest_widget = GUIEngine::getWidget(next_id);
+        closest_widget->setFocusForPlayer(playerID);
+
+        // A list exception : when entering a list by going down/left/right, select the first item
+        // when focusing a list by going up, select the last item of the list
+        if (closest_widget->m_type == WTYPE_LIST)
         {
-            IGUIElement *closest = GUIEngine::getGUIEnv()->getRootGUIElement()->getElementFromId(el->getTabOrder() + (reverse ? -n : n), true);
+            ListWidget* list = (ListWidget*) closest_widget;
+            assert(list != NULL);
+            list->setSelectionID(nav == NAV_UP ? list->getItemCount() - 1 : 0);
+        }
+    }
 
-            if (closest != NULL && Widget::isFocusableId(closest->getID()))
-            {
-                Widget* closestWidget = GUIEngine::getWidget( closest->getID() );
+    return;
 
-                if (playerID != PLAYER_ID_GAME_MASTER && !closestWidget->m_supports_multiplayer) return;
+} // navigate
 
-                // if a dialog is shown, restrict to items in the dialog
-                if (ScreenKeyboard::isActive())
-                {
-                    if (!ScreenKeyboard::getCurrent()->isMyChild(closestWidget))
-                        continue;
-                }
-                else if (ModalDialog::isADialogActive())
-                {
-                    if (!ModalDialog::getCurrent()->isMyChild(closestWidget))
-                        continue;
-                }
+/* This function use simple heuristic to find the closest widget
+ * in the requested direction,
+ * It prioritize widgets close vertically to widget close horizontally,
+ * as it is expected behavior in any direction.
+ * Several hardcoded values are used, having been found to work well
+ * experimentally while keeping simple heuristics.
+ */
+int EventHandler::findIDClosestWidget(const NavigationDirection nav, const int playerID)
+{
+    Widget* w = GUIEngine::getFocusForPlayer(playerID);
 
-                if (NAVIGATION_DEBUG)
-                {
-                    Log::info("EventHandler", "Navigating %s to %d", (reverse ? "up" : "down"), closest->getID());
-                }
+    int closest_widget_id = -1;
+    int distance = 0;
+    // So that the UI behavior don't change when it is upscaled
+    const int BIG_DISTANCE = irr_driver->getActualScreenSize().Width*10;
+    int smallest_distance = BIG_DISTANCE;
+    // Used when there is no suitable widget in the requested direction
+    int closest_wrapping_widget_id = -1;
+    int wrapping_distance = 0;
+    int smallest_wrapping_distance = BIG_DISTANCE;
+
+    // In theory, it's better to look recursively in m_widgets
+    // In practice, this is much much simpler and work equally well
+    // Usual widget IDs begin at 100 and there is rarely more
+    // than a few dozen of them - but it's very cheap to have some margin,
+    for (int i=0;i<1000;i++)
```

Comment:
This is a pretty dirty hack

Reply:
This is not great from the plan of theory and idealized conceptual computer science, where the time needed to write the perfectly elegant program is not factored in.

But from the engineering point-of-view, it does the job well, it is easier to maintain that some recursive-access widget selection, and it saves me the hours needed to find, write and debug the clumsy chain of OOP objects needed to get the data (the mixing of data and functions in OOP is very impractical).